File: go/0046M-全排列.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-15 09:03:11
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func permute(nums []int) [][]int {
	ans := make([][]int, 0)
	per := make([]int, 0)
	vis := make([]bool, len(nums))
	var dfs func(index int)
	dfs = func(index int) {
		if len(per) == len(nums) {
			tmp := make([]int, len(per))
			copy(tmp, per)
			ans = append(ans, tmp)
			return
		}
		for i := 0; i < len(nums); i++ {
			if vis[i] {
				continue
			}
			per = append(per, nums[i])
			vis[i] = true
			dfs(i + 1)
			vis[i] = false
			per = per[:len(per)-1]
		}
	}
	dfs(0)
	return ans
}

File: cpp/0046M-全排列.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-27 16:36:56
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;

    void dfs(vector<int>& nums, vector<bool>& vis) {
        if (tmp.size() == nums.size()) {
            ans.push_back(tmp);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (vis[i]) continue;
            tmp.push_back(nums[i]);
            vis[i] = true;
            dfs(nums, vis);
            vis[i] = false;
            tmp.pop_back();
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> vis(nums.size(), false);
        dfs(nums, vis);
        return ans;
    }
};
